circom/client_side_proofs/oprf_query.circom

```diff
@@ -8,53 +8,132 @@ include "babyjubjub/babyjubjub.circom";
 
 // Checks outside of the ZK proof: The output point q needs to be a valid BabyJubJub point in the correct subgroup.
 
-template OprfQuery(MAX_DEPTH) {
+// The MerkleTree has 7 public keys at its leaf. We need to hash them to proof membership
+template MerkleLeaf() {
+    signal input pk[7][2];
+    signal output out;
+
+    component hasher = Poseidon2(16);
+    hasher.in[0] <== 0;
+    for (var i = 0; i < 7; i++) {
+        hasher.in[i * 2 + 1] <== pk[i][0];
+        hasher.in[i * 2 + 2] <== pk[i][1];
+    }
+    hasher.in[15] <== 0;
+    out <== hasher.out[1];
+}
+
+// Choses a public key from the list of NUM_KEYS public keys based on the index. If the index is out of range, the result is zero. There is no check enforcing this zero, since in this use case it is checked later on anyway.
+template ChoosePublicKey(NUM_KEYS) {
+    signal input pk[NUM_KEYS][2];
+    signal input index; // 0..NUM_KEYS-1
+    signal output out[2];
+
+    component cmp[NUM_KEYS];
+    for (var i = 0; i < NUM_KEYS; i++) {
+        cmp[i] = IsEqual();
+        cmp[i].in[0] <== index;
+        cmp[i].in[1] <== i;
+    }
+
+    // We know that cmp is a vector with at most one 1 (else is 0), so we can do a CMUX chain.
+    signal cmux[NUM_KEYS][2];
+    cmux[0][0] <== cmp[0].out * pk[0][0]; // Initialize with 0 or the first public key
+    cmux[0][1] <== cmp[0].out * pk[0][1]; // Initialize with 0 or the first public key
+    for (var i = 1; i < NUM_KEYS; i++) {
+        cmux[i][0] <== cmux[i - 1][0] + cmp[i].out * (pk[i][0] - cmux[i - 1][0]);
+        cmux[i][1] <== cmux[i - 1][1] + cmp[i].out * (pk[i][1] - cmux[i - 1][1]);
+    }
```

Comment:
If the CMP outputs are just a one-hot vector, can this not just be a dot-product?